File: lena-config/src/test/java/org/riversoforion/lena/config/PrioritizedConfigurationSourceTest.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class PrioritizedConfigurationSourceTest {

    @Test
    @DisplayName("getValue from a single source")
    void getValue_SingleSource(@Mock ConfigurationSource source) {

        Namespace ns = Namespace.of("prefix");
        when(source.getValue(eq(ns), any())).thenReturn(Optional.empty());
        when(source.getValue(eq(ns), startsWith("existing"))).thenAnswer(invocation -> {
            String name = invocation.getArgument(1, String.class);
            return Optional.of(name + " value");
        });

        PrioritizedConfigurationSource prioritized = new PrioritizedConfigurationSource(List.of(source));

        assertThat(prioritized.getValue(ns, "existing num")).isPresent()
                                                            .contains("existing num value");
        assertThat(prioritized.getValue(ns, "existing str")).isPresent()
                                                            .contains("existing str value");
        assertThat(prioritized.getValue(ns, "missing num")).isEmpty();
    }

    @Test
    @DisplayName("getValue from multiple sources")
    void getValue_MultipleSources(@Mock ConfigurationSource first, @Mock ConfigurationSource second, @Mock ConfigurationSource third) {

        Namespace ns = Namespace.of("this", "that");
        // By default, value is "missing"
        when(first.getValue(eq(ns), any())).thenReturn(Optional.empty());
        when(second.getValue(eq(ns), any())).thenReturn(Optional.empty());
        when(third.getValue(eq(ns), any())).thenReturn(Optional.empty());
        // Test scenarios
        when(first.getValue(ns, "first num")).thenReturn(Optional.of("first value"));
        when(second.getValue(ns, "second bool")).thenReturn(Optional.of("second value"));
        when(third.getValue(ns, "third string")).thenReturn(Optional.of("third value"));

        PrioritizedConfigurationSource prioritized = new PrioritizedConfigurationSource(List.of(first, second, third));

        assertThat(prioritized.getValue(ns, "first num")).isPresent()
                                                         .contains("first value");
        assertThat(prioritized.getValue(ns, "second bool")).isPresent()
                                                           .contains("second value");
        assertThat(prioritized.getValue(ns, "third string")).isPresent()
                                                            .contains("third value");
        assertThat(prioritized.getValue(ns, "fourth num")).isEmpty();

        verify(first, times(4)).getValue(eq(ns), anyString());
        verify(second, times(3)).getValue(eq(ns), anyString());
        verify(third, times(2)).getValue(eq(ns), anyString());
    }
}

File: lena-config/src/commonTest/kotlin/org/riversoforion/lena/config/PrioritizedConfigurationSourceTest.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class PrioritizedConfigurationSourceTest {

    @Test
    fun getValue_SingleSource() {
        val ns = Namespace.of("prefix")
        val source = ConfigurationSource { namespace, name ->
            if (namespace != ns) null
            else if (name.segments().first().startsWith("existing")) "$name value" else null
        }
        val prioritized = PrioritizedConfigurationSource(listOf(source to null))

        assertEquals("existing-num value", prioritized.getValue(ns, Name.of("existing num")))
        assertEquals("existing-str value", prioritized.getValue(ns, Name.of("existing str")))
        assertNull(prioritized.getValue(ns, Name.of("missing num")))
    }

    @Test
    fun getValue_MultipleSources() {
        val ns = Namespace.of("this", "that")
        var firstCalls = 0
        var secondCalls = 0
        var thirdCalls = 0
        val first = ConfigurationSource { namespace, name ->
            firstCalls++
            if (namespace == ns && name == Name.of("first num")) "first value" else null
        }
        val second = ConfigurationSource { namespace, name ->
            secondCalls++
            if (namespace == ns && name == Name.of("second bool")) "second value" else null
        }
        val third = ConfigurationSource { namespace, name ->
            thirdCalls++
            if (namespace == ns && name == Name.of("third string")) "third value" else null
        }
        val prioritized = PrioritizedConfigurationSource(
            listOf(first to null, second to null, third to null)
        )

        assertEquals("first value", prioritized.getValue(ns, Name.of("first num")))
        assertEquals("second value", prioritized.getValue(ns, Name.of("second bool")))
        assertEquals("third value", prioritized.getValue(ns, Name.of("third string")))
        assertNull(prioritized.getValue(ns, Name.of("fourth num")))

        assertEquals(4, firstCalls)
        assertEquals(3, secondCalls)
        assertEquals(2, thirdCalls)
    }
}
